sympy/tensor/array/array_derivatives.py

```diff
@@ -0,0 +1,121 @@
+from sympy import Derivative, Integer, Expr
+from sympy.matrices.common import MatrixCommon
+from .ndim_array import NDimArray
+from .arrayop import derive_by_array
+from sympy import MatrixExpr
+from sympy import ZeroMatrix
+from sympy.matrices.expressions.matexpr import _matrix_derivative
+
+
+class ArrayDerivative(Derivative):
+
+    is_scalar = False
+
+    def __new__(cls, expr, *variables, **kwargs):
+        obj = super(ArrayDerivative, cls).__new__(cls, expr, *variables, **kwargs)
+        if isinstance(obj, ArrayDerivative):
+            obj._shape = obj._get_shape()
```

Comment:
There is a `cached_property` decorator in ode.py for this sort of thing:
https://github.com/sympy/sympy/blob/b6c328e2f057ba81513d19a1451ace358ecc8233/sympy/solvers/ode/single.py#L36-L45

Perhaps that could be moved to `sympy/core/decorators.py` and then reused for `shape` here.

Reply:
Well, concatenating a few tuples may be computationally less expensive than caching a property.

Reply:
I will be surprised if you can measure a difference in any real calculation.

The class here is already using a cached property. The shape property returns this `_shape` attribute which is precomputed in the constructor. The `cached_property` decorator exists to make this pattern as simple as applying a decorator to the property method and also has the advantage of not computing the result if it is never accessed. That's a big performance improvement in some other situations although likely insignificant in this case.

Reply:
OK, maybe in a separate PR.